### PR TITLE
Task checkpointing #234

### LINKED ISSUE #234: Race condition in task checkpointing
With `config["globals"]["checkpointMode"] = "task_exit"`, the DFK attempts to checkpoint every task at task completion. However in the case of fast apps without dependencies, the task returns before the AppFuture is created. This results in a race condition between the checkpoint method and the task update.

If the checkpoint method is called first we get :
```
Traceback (most recent call last):
  File "test_regression_233.py", line 38, in <module>
    test_checkpoint_availability()
  File "test_regression_233.py", line 27, in test_checkpoint_availability
    original = run_checkpointed([])
  File "test_regression_233.py", line 18, in run_checkpointed
    x = cached_rand(i)
  File "/home/yadu/src/parsl/parsl/app/app_factory.py", line 76, in __call__
    return app_obj(*args, **kwargs)
  File "/home/yadu/src/parsl/parsl/app/python_app.py", line 46, in __call__
    **kwargs)
  File "/home/yadu/src/parsl/parsl/dataflow/dflow.py", line 499, in submit
    task_id, func, *new_args, **kwargs)
  File "/home/yadu/src/parsl/parsl/dataflow/dflow.py", line 322, in launch_task
    exec_fu.add_done_callback(partial(self.handle_update, task_id))
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 376, in add_done_callback
    fn(self)
  File "/home/yadu/src/parsl/parsl/dataflow/dflow.py", line 218, in handle_update
    self.checkpoint(tasks=[task_id])
  File "/home/yadu/src/parsl/parsl/dataflow/dflow.py", line 617, in checkpoint
    if self.tasks[task_id]['app_fu'].done() and \
AttributeError: 'NoneType' object has no attribute 'done'
```

### PR DESCRIPTION
**Pull request by yadudoc**
Fixes #234 

DFK relies on internal task status instead of futures, avoiding a race-condition.
Updates the task_exit test to look at the checkpoint files rather than logs.

### CODE REVIEW

File: parsl/dataflow/dflow.py [hunk 1]
```diff
@@ -606,10 +621,13 @@ def checkpoint(self):
...
         count = 0
+
         with open(checkpoint_tasks, 'ab') as f:
-            for task_id in self.tasks:
-                if self.tasks[task_id]['app_fu'].done() and \
-                   not self.tasks[task_id]['checkpoint']:
+            for task_id in checkpoint_queue:
+
+                if not self.tasks[task_id]['checkpoint'] and \
+                   self.tasks[task_id]['status'] in (States.done, States.failed, States.dep_fail):
```
Comment (annawoodard) [inline parsl/dataflow/dflow.py]: I don't quite follow why we would checkpoint failed tasks? For example, what if there was a transient issue causing tasks to fail, the user stops their script and fixes the issue, then re-runs their script?

File: parsl/tests/test_cleanup/test_cleanup.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        f = slow_double(i)
+        round_1.append(f)
+
+    round_2 = []
+    for i in round_1:
+        f = slow_double(i)
+        round_2.append(f)
+
+    dfk.cleanup()
+
```
Comment (annawoodard) [inline parsl/tests/test_cleanup/test_cleanup.py]: I think we need to assert something here?

File: parsl/dataflow/dflow.py (hunk 1 above)
Comment (benclifford) [reply parsl/dataflow/dflow.py]: I think checkpointing failures is the wrong thing to do (in most cases?)

File: parsl/tests/test_cleanup/test_cleanup.py (hunk 2 above)
Comment (yadudoc) [reply parsl/tests/test_cleanup/test_cleanup.py]: Well, I don't know what to assert yet. I think the mistake was in this test getting added to this PR.
This was for figuring out the #221 behavior.

File: parsl/dataflow/dflow.py (hunk 1 above)
Comment (yadudoc) [reply parsl/dataflow/dflow.py]: Makes sense. There's so much external, non-trackable things that will influence apps.

I'll change this behavior.

File: parsl/tests/test_cleanup/test_cleanup.py (hunk 2 above)
Comment (yadudoc) [reply parsl/tests/test_cleanup/test_cleanup.py]: This (incomplete) test should've gone in on a separate PR. I was watching for cleanup behavior, and there isn't something to assert on here yet.

File: parsl/dataflow/dflow.py [hunk 3]
```diff
@@ -572,12 +578,14 @@ def cleanup(self):
...
 
-    def checkpoint(self):
+    def checkpoint(self, tasks=None):
         """Checkpoint the dfk incrementally to a checkpoint file.
 
         When called, every task that has been completed yet not
         checkpointed is checkpointed to a file.
 
+        Kwargs:
+            - tasks (List of task ids) : List of task ids to checkpoint. Default=None
```
Comment (benclifford) [inline parsl/dataflow/dflow.py]: clarify what happens in None case (assuming that the intention isn't to checkpoint nothing...)

File: parsl/tests/test_sites/test_ec2/ec2.py [hunk 4]
```diff
@@ -1,91 +0,0 @@
-ec2OnDemand = {
```
Comment (benclifford) [inline parsl/tests/test_sites/test_ec2/ec2.py]: is this EC2 stuff relevant to the PR?

Comment (yadudoc) [reply parsl/tests/test_sites/test_ec2/ec2.py]: It's not I keep pushing these changes to the wrong branch instead of master. While not related, can we keep it so that it'll get to master ?

File: parsl/tests/test_checkpointing/test_regression_239.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            while f:
+                tasks.append(pickle.load(f))
+                print
+        except EOFError:
+            pass
+        print("Tasks from cache : ", tasks)
+        assert len(tasks) == 1, "Expected {} checkpoint events, got {}".format(1, len(tasks))
+
+
+def test_checkpointing_at_dfk_exit():
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/tests/test_checkpointing/test_regression_239.py]: Agreed. Removing dfk_exit test.

File: parsl/tests/test_checkpointing/test_regression_239.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+from parsl import App, DataFlowKernel
+import pickle
+import time
+
+
+def run_checkpointed(n=5, mode="task_exit"):
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/tests/test_checkpointing/test_regression_239.py]: Done

File: parsl/dataflow/dflow.py [hunk 7]
```diff
@@ -573,12 +578,16 @@ def cleanup(self):
 
         logger.info("DFK cleanup complete")
 
-    def checkpoint(self):
+    def checkpoint(self, tasks=None):
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/dataflow/dflow.py]: With checkpointing mode `task_exit` the checkpoint method is called at the completion of each task. Earlier, we were iterating over the entire list of tasks to identify what is to be checkpointed, when we really only needed to checkpoint a specific task.
